text/0000-variant-types.md

```diff
@@ -0,0 +1,243 @@
+let b: Foo = a; // Ok
+let _: Foo::Variant2 = a; // Compile-time error
+let _: Foo::Variant2 = b; // Compile-time error
+let _ = a as Foo::Variant2; // Compile-time error
+let _ = b as Foo::Variant2; // Runtime error
+let _ = b as Foo::Variant1; // Ok
+```
+
+See alternatives below, it may be better to not support down-casting.
+
+
+## impls
+
+`impl`s may exist for both enum and variant types. There is no explicit sharing
+of impls, and just because as enum has a trait bound, does not imply that the
+variant also has that bound. However, the usual conversion rules apply, so if a
+method would apply to the enum type, it can be called on a variant value due to
+coercion performed by the dot operator.
+
+
+## Extension - unsized enums
+
+With the above representation, enum variants are the same size as the enum
+itself, which is the size of the largest enum plus the discriminant. This makes
+conversion between the variant and enum types easy, and should be the default.
+However, there are some use cases where it is preferable to have a more minimal
+size for variant values. For example, where variants are of wildly different
+sizes and where we usually deal with individual variants and rarely the whole
+enum.
+
+For such use cases, we could support an `#[unsized]` attribute on the enum. This
+affects the representation of the variants: a variant value is not padded, it
+still has the discriminant, but there is no padding to the enum size, the value
+is the size of the individual variant (plus the discriminant, of course).
+
+The enum type (but not the variant types) are considered unsized. The effect is
+that they may not appear in a Rust program by value, only by reference. There is
+no 'unsizing information' (c.f., slices or trait objects) so a pointer to an
+unsized enum is a regular pointer, not a fat pointer.
+
+Casting/coercion of enum/variant values can still work as before, since we'll
+never access the enum and find a different variant.
+
+### Further extension - remove the discriminant
+
+We don't need the discriminant when we have a value with variant type. We can't
+have a value with enum type. When we have a pointer with enum type, we could put
+the discriminant in the pointer (making it a fat pointer like we use with other
+unsized types). This should all work at the expense of some added complexity in
+coercion and matching.
+
+
+# Drawbacks
+
+The proposal is a little bit hairy, in part due to trying to remain backwards
+compatible.
+
+
+# Alternatives
+
+An alternative to allowing variants as types is allowing sets of variants as
+types, a kind of refinement type. This set could have one member and then would
+be equivalent to variant types, or could have all variants as members, making it
+equivalent to the enum type. Although more powerful, this approach is more
+complex, and I do not believe the complexity is justified.
+
+We could remove support for casting from enums to variants, relying on matching.
```

Comment:
casting can always be added later, and it would be the first cast that can panic at runtime.

Also it's trivial to write an enum downcast macro that panics or does something other user defined.